Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp

```diff
@@ -1068,11 +1069,22 @@ void MediaPlayerPrivateGStreamerMSE::attemptToDecryptWithInstance(const CDMInsta
         //  ClearKey for some reason.
         auto& cdmInstanceOpenCDM = downcast<CDMInstanceOpenCDM>(instance);
         GST_TRACE("instance is OpenCDM, continuing with %s", cdmInstanceOpenCDM.keySystem().utf8().data());
-        String sessionId = cdmInstanceOpenCDM.getCurrentSessionId();
-        ASSERT(!sessionId.isEmpty());
-        GUniquePtr<GstStructure> structure(gst_structure_new("drm-session", "session", G_TYPE_STRING, sessionId.utf8().data(), nullptr));
-        for (const auto& it : m_appendPipelinesMap)
-            it.value->dispatchDecryptionStructure(GUniquePtr<GstStructure>(gst_structure_copy(structure.get())));
+
+        for (const auto& it : m_appendPipelinesMap) {
+            if (!(it.value->isDecryptionStructureDispatched())) {
+                String sessionId = cdmInstanceOpenCDM.getSessionIdByInitData(SharedBuffer::create(it.value->getInitData()));
+                // FIXME: If initData is not mapped, then retrieve the current sessionId in case of single initData.
+                if (sessionId.isEmpty()) {
```

Comment:
Same case as for regular playback, let's optimize the search.

Reply:
will change the order and FIXME comment as per regular flow comment

Reply:
Please, reconsider the whole MSE flow taking into account the considerations I wrote for the normal flow. Now my head is going to explode if I think of this flow anymore.

Reply:
Updated